### PR TITLE
Add missing flag to manager in sample provider code

### LINKED ISSUE #742: CrashLoopBackOff executing provider following book instructions
/kind bug

When developing a cluster-api provider [according to the instructions in the cluster-api book](https://kubernetes-sigs.github.io/cluster-api/provider_implementations/overview.html), the pod for the manager enters in a `CrashLoopBackOff` status with the following error:
```
kubectl logs cluster-api-provider-solas-controller-manager-0 --namespace cluster-api-provider-solas-system -c manager
flag provided but not defined: -metrics-addr
Usage of /manager:
  -kubeconfig string
    	Paths to a kubeconfig. Only required if out-of-cluster.
  -master string
    	The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.
```

Comparing [the code provided in the book](https://kubernetes-sigs.github.io/cluster-api/provider_implementations/register_schemes.html) for  `cmd/manager/main.go` with he code automatically generated by kubebuilder (see [this example](https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/project/cmd/manager/main.go)) we found that the definition for the `metrics-addr` is missing:
```
var metricsAddr string
 flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
```
This is problematic because the manifest for the provider's pod passes this argument:
```
  - args:
    - --metrics-addr=127.0.0.1:8080
    command:
    - /manager
    env:
    - name: POD_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: SECRET_NAME
      value: cluster-api-provider-solas-webhook-server-secret
    image: cluster-api-provider-solas
    imagePullPolicy: IfNotPresent
    name: manager
``` 

The addition of this argument can be traced to the patch at `config/default/manager_auth_proxy_patch.yaml` applied during customization:
```
      - name: manager 
        args:
        - "--metrics-addr=127.0.0.1:8080"
``` 
This patch is referenced in `config/default/kustomization.yaml' 
```
 # Comment the following 3 lines if you want to disable
  # the auth proxy (https://github.com/brancz/kube-rbac-proxy)
  # which protects your /metrics endpoint.
- ../rbac/auth_proxy_service.yaml
- ../rbac/auth_proxy_role.yaml
- ../rbac/auth_proxy_role_binding.yaml
```
**Environment:**

* kubebuilder 1.0.8
* Cluster-api version: 0.0.0-alpha.4
* Kubernetes version: (use `kubectl version`): v1.13.2
* OS (e.g. from `/etc/os-release`): OpenSUSE Leap 15.0

### PR DESCRIPTION
**Pull request by pablochacin**
Fix #742 CrashLoopBackOff in sample provider from book due to missing flag in manager

### CODE REVIEW

File: docs/book/provider_implementations/register_schemes.md [hunk 1]
```diff
@@ -48,13 +48,16 @@ func main() {
                 panic(fmt.Errorf("GetConfigOrDie didn't die"))
         }
 
+        var metricsAddr string
+        flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
```
Comment (vincepri) [inline docs/book/provider_implementations/register_schemes.md]:
```suggestion
        var metricsAddr = flag.String(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
```

outside the main function

Comment (pablochacin) [reply docs/book/provider_implementations/register_schemes.md]: I suppose you meant:
```
        var metricsAddr = flag.String("metrics-addr", ":8080", "The address the metric endpoint binds to.")
```

**Comment by gyliu513** [comment]
@pablochacin did you test this PR? I found build failed as follows:

```
root@gyliu-dev1:~/go/src/sigs.k8s.io/cluster-api-provider-genesis# make
go generate ./pkg/... ./cmd/...
go fmt ./pkg/... ./cmd/...
go vet ./pkg/... ./cmd/...
# sigs.k8s.io/cluster-api-provider-genesis/cmd/manager [sigs.k8s.io/cluster-api-provider-genesis/cmd/manager.test]
cmd/manager/main.go:51:47: cannot use metricsAddr (type *string) as type string in field value
Makefile:40: recipe for target 'vet' failed
make: *** [vet] Error 2
```

File: docs/book/provider_implementations/register_schemes.md [hunk 2]
```diff
@@ -48,13 +48,15 @@ func main() {
...
+        metricsAddr := flag.String("metrics-addr", ":8080", "The address the metric endpoint binds to.")
         flag.Parse()
+
         log := logf.Log.WithName("solas-controller-manager")
         logf.SetLogger(logf.ZapLogger(false))
         entryLog := log.WithName("entrypoint")
 
         // Setup a Manager
-        mgr, err := manager.New(cfg, manager.Options{})
+        mgr, err := manager.New(cfg, manager.Options{MetricsBindAddress: metricsAddr})
```
Comment (detiber) [inline docs/book/provider_implementations/register_schemes.md]:
```suggestion
        mgr, err := manager.New(cfg, manager.Options{MetricsBindAddress: *metricsAddr})
```

File: docs/book/provider_implementations/register_schemes.md [hunk 3]
```diff
@@ -48,13 +48,15 @@ func main() {
                 panic(fmt.Errorf("GetConfigOrDie didn't die"))
         }
 
+        metricsAddr := flag.String("metrics-addr", ":8080", "The address the metric endpoint binds to.")
```
Comment (vincepri) [inline docs/book/provider_implementations/register_schemes.md]: Go flag variables should be defined outside the main function

**Comment by pablochacin** [comment]
> @pablochacin did you test this PR? I found build failed as follows with your PR:

No, I'm sorry. I had tested the previous version. Thanks @detiber for chasing the error.

File: docs/book/provider_implementations/register_schemes.md (hunk 3 above)
Comment (pablochacin) [reply docs/book/provider_implementations/register_schemes.md]: I've seen many examples not doing so. Is this a convention for cluster-api?

Comment (davidewatson) [reply docs/book/provider_implementations/register_schemes.md]: In general I agree with @vincepri here (making the variable global in scope make it easier to reuse as a library). In this case I wonder if we should attempt to follow `kubebuilder` instead? https://github.com/cattlek8s/cluster-api-provider-generic/blob/master/cmd/manager/main.go#L38

Comment (gyliu513) [reply docs/book/provider_implementations/register_schemes.md]: @davidewatson but the example here 
https://github.com/cattlek8s/cluster-api-provider-generic/blob/master/cmd/manager/main.go#L38 is using flag inside main function, so do you mean we need to keep it in main or move it out of main?

File: docs/book/provider_implementations/register_schemes.md [hunk 4]
```diff
@@ -48,13 +48,15 @@ func main() {
...
         flag.Parse()
+
         log := logf.Log.WithName("solas-controller-manager")
         logf.SetLogger(logf.ZapLogger(false))
         entryLog := log.WithName("entrypoint")
 
         // Setup a Manager
-        mgr, err := manager.New(cfg, manager.Options{})
+        mgr, err := manager.New(cfg, manager.Options{MetricsBindAddress: *metricsAddr})
```
Comment (gyliu513) [inline docs/book/provider_implementations/register_schemes.md]: Thanks @pablochacin , this works ;-)

File: docs/book/provider_implementations/register_schemes.md (hunk 3 above)
Comment (davidewatson) [reply docs/book/provider_implementations/register_schemes.md]: I meant that defining the variable within the function is consistent with what `kubebuilder` does.

**Comment by ncdc** [comment]
```
I0306 20:28:06.838] === RUN   TestReconcile
I0306 20:28:06.838] E0306 20:27:58.726606   20132 controller.go:375] Unable to retrieve Machineset default/foo-f599856b8 from store: MachineSet.cluster.k8s.io "foo-f599856b8" not found
I0306 20:28:06.839] E0306 20:28:01.678876   20132 controller.go:179] Failed to reconcile MachineDeployment "default/foo": Operation cannot be fulfilled on machinesets.cluster.k8s.io "foo-74c978d5b5": the object has been modified; please apply your changes to the latest version and try again
I0306 20:28:06.839] --- FAIL: TestReconcile (8.14s)
I0306 20:28:06.839]     machinedeployment_controller_test.go:200: error updating machineset: Operation cannot be fulfilled on machinesets.cluster.k8s.io "foo-74c978d5b5": the object has been modified; please apply your changes to the latest version and try again
I0306 20:28:06.840]     machinedeployment_controller_test.go:202: timed out waiting reconcile error
I0306 20:28:06.840] FAIL
I0306 20:28:06.840] FAIL	sigs.k8s.io/cluster-api/pkg/controller/machinedeployment	37.136s
```

May be a bug?

/retest

**Pull request merged**